### PR TITLE
Optimizing LCO continuations

### PR DESCRIPTION
**Pull request by sithhell**
Instead of routing set_value/set_error/trigger_event actions to the
destination, we send the lva address along with the continuation to
avoid further AGAS lookups

### CODE REVIEW

File: hpx/lcos/promise.hpp [hunk 1]
```diff
@@ -532,7 +532,17 @@ namespace hpx { namespace lcos
...
         }
 
-        /// \brief Return the global id of this \a future instance
+        /// \brief Return the resolved address of this \a promise instance
+        naming::address resolve() const
+        {
+            return naming::address(
+                hpx::get_locality()
+              , impl_->get_component_type()
+              , &*impl_
```
Comment (hkaiser) [inline hpx/lcos/promise.hpp]: I'd prefer writing this as `impl_.get()`.

File: hpx/lcos/promise.hpp [hunk 2]
```diff
@@ -676,7 +686,17 @@ namespace hpx { namespace lcos
...
```
Comment (hkaiser) [inline hpx/lcos/promise.hpp]: Same as above.

File: hpx/runtime/applier/detail/apply_implementations.hpp [hunk 3]
```diff
@@ -54,6 +54,48 @@ namespace hpx { namespace detail
...
+                ) % hpx::actions::detail::get_action_name<Action>()));
+            return false;
+        }
+
+        // Determine whether the id is local or remote
+        if(addr)
+        {
+            if(addr.locality_ == hpx::get_locality())
+            {
+                std::pair<bool, components::pinned_ptr> r;
```
Comment (hkaiser) [inline hpx/runtime/applier/detail/apply_implementations.hpp]: Shouldn't this definition be moved out of the `if() {}` block? Also, wouldn't the trait be called twice if it returns false the first time?

File: hpx/runtime/applier/detail/apply_implementations.hpp [hunk 4]
```diff
@@ -87,6 +129,35 @@ namespace hpx { namespace detail
...
+    bool apply_impl(hpx::id_type const& id, naming::address addr,
+        threads::thread_priority priority, Ts &&... vs)
+    {
+        // Determine whether the id is local or remote
+        if(addr)
+        {
+            if(addr.locality_ == hpx::get_locality())
+            {
+                std::pair<bool, components::pinned_ptr> r;
+                r = traits::action_was_object_migrated<Action>::call(
```
Comment (hkaiser) [inline hpx/runtime/applier/detail/apply_implementations.hpp]: Same as above.

File: hpx/runtime/applier/trigger.hpp [hunk 5]
```diff
@@ -10,6 +10,7 @@
 
 #include <hpx/runtime/actions/continuation.hpp>
 
+/*
```
Comment (hkaiser) [inline hpx/runtime/applier/trigger.hpp]: Was this removed intentionally? Or is it an oversight?

File: tests/regressions/lcos/promise_1620.cpp [hunk 6]
```diff
@@ -70,7 +70,7 @@ int hpx_main(int argc, char* argv[])
 
         // This segfaults, because the promise is not alive any more.
         // It SHOULD get kept alive by AGAS though.
-        hpx::set_lco_value(promise_id, 10, false);
+        hpx::set_lco_value(promise_id, naming::address(), 10, false);
```
Comment (hkaiser) [inline tests/regressions/lcos/promise_1620.cpp]: Would it be possible to leave the old API in place?

File: tests/unit/agas/components/server/managed_refcnt_checker.cpp [hunk 7]
```diff
@@ -47,7 +47,7 @@ managed_refcnt_checker::~managed_refcnt_checker()
         strm << ( boost::format("[%1%/%2%]: triggering flag %3%\n")
                 % prefix_ % this_ % target_);
 
-        hpx::trigger_lco_event(target_);
+        hpx::trigger_lco_event(target_, naming::address());
     }
```
Comment (hkaiser) [inline tests/unit/agas/components/server/managed_refcnt_checker.cpp]: I'd like to leave the old API in place here as well. Having to pass a default constructed address() is confusing.

File: hpx/runtime/applier/trigger.hpp (hunk 5 above)
Comment (sithhell) [inline hpx/runtime/applier/trigger.hpp]: Yes, this was removed intentionally. the whole file could go as those functions aren't used anywhere.

File: hpx/runtime/applier/detail/apply_implementations.hpp (hunk 3 above)
Comment (sithhell) [inline hpx/runtime/applier/detail/apply_implementations.hpp]: You are right, the definition has to be removed out of the if block.
The trait has to go inside the if block though, i guess.

File: tests/regressions/lcos/promise_1620.cpp (hunk 6 above)
Comment (sithhell) [inline tests/regressions/lcos/promise_1620.cpp]: Ok, as you wish.

File: hpx/runtime/applier/trigger.hpp (hunk 5 above)
Comment (hkaiser) [inline hpx/runtime/applier/trigger.hpp]: Those are API functions, I'd say we should leave them in. They are useful in their own right.

**Comment by sithhell** [comment]
The review comments have been addressed.

**Comment by hkaiser** [comment]
[10 Feb 16 06:40] * hkaiser * heller: couldn't the continuations take the address by &&?
[10 Feb 16 06:41] * heller * hkaiser: most likely they could indeed
[10 Feb 16 06:43] * heller * such that we move the address out of the continuation?
[10 Feb 16 06:43] * hkaiser * well, we know it's used only once, don't we?
[10 Feb 16 06:43] * heller * i think so, yes

**Comment by sithhell** [comment]
Do you want to see that implemented right away or should we rather merge it now (after circle ci passes) and create a ticket about fixing this so we implement it later on?

**Comment by sithhell** [comment]
The reason why I ask is because we could additionally do the same for the GID.

**Comment by hkaiser** [comment]
>> Isn't this dangerous? Different compilers might use different packing schemes for the consecutive data members...
>
> Good point, don't we have some guarantee for that? Also, what's the difference to `name_serialization_data` defined in address.cpp or gid_type?

You're right, we probably never have attempted to pay attention to this issue. The same problem applies to hpx::util::tuple<Ts...>, (for Ts being bitwise serializable)...

**Comment by hkaiser** [comment]
> Do you want to see that implemented right away or should we rather merge it now (after circle ci passes) and create a ticket about fixing this so we implement it later on?

We move `naming::address` everywhere, so I'd say let's fix this. Passing id_type by rvalue does not buy us too much, so this can wait.

**Comment by sithhell** [comment]
> We move naming::address everywhere, so I'd say let's fix this. Passing id_type by rvalue does not buy us too much, so this can wait.

done.

File: tests/unit/parcelset/set_parcel_write_handler.cpp [hunk 8]
```diff
@@ -12,7 +12,7 @@
 ///////////////////////////////////////////////////////////////////////////////
 void test(hpx::id_type const& id)
 {
-    hpx::trigger_lco_event(id);
+    hpx::trigger_lco_event(id, hpx::naming::address());
```
Comment (hkaiser) [inline tests/unit/parcelset/set_parcel_write_handler.cpp]: Minor nitpick: this change is not necessary anymore.

**Comment by hkaiser** [comment]
LGTM! Please go ahead and merge.

**Pull request merged**